File: src/Java/myservicebus/src/main/java/com/myservicebus/persistence/TransportOutboxDispatcher.java
package com.myservicebus.persistence;

import com.myservicebus.BusHook;
import com.myservicebus.MessageOperationHookEvent;
import com.myservicebus.SendTransport;
import com.myservicebus.TransportFactory;
import com.myservicebus.tasks.CancellationToken;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;

public final class TransportOutboxDispatcher implements OutboxTransportDispatcher {
    private final TransportFactory transportFactory;
    private final List<BusHook> hooks;

    public TransportOutboxDispatcher(TransportFactory transportFactory) {
        this(transportFactory, List.of());
    }

    public TransportOutboxDispatcher(TransportFactory transportFactory, Iterable<? extends BusHook> hooks) {
        this.transportFactory = Objects.requireNonNull(transportFactory, "transportFactory");
        Objects.requireNonNull(hooks, "hooks");
        java.util.ArrayList<BusHook> copiedHooks = new java.util.ArrayList<>();
        hooks.forEach(copiedHooks::add);
        this.hooks = List.copyOf(copiedHooks);
    }

    @Override
    public CompletableFuture<Void> dispatch(OutboxMessage message, CancellationToken cancellationToken) {
        Objects.requireNonNull(message, "message");
        Objects.requireNonNull(cancellationToken, "cancellationToken");
        if (cancellationToken.isCancelled()) {
            return CompletableFuture.failedFuture(
                    new java.util.concurrent.CancellationException("Outbox dispatch was cancelled."));
        }

        long startedAt = System.nanoTime();
        try {
            SendTransport transport = transportFactory.getSendTransport(message.destinationAddress());
            Map<String, Object> headers = new LinkedHashMap<>(message.headers());
            headers.put("_content_type", message.contentType());
            headers.put("_message_id", message.messageId().toString());
            if (message.correlationId() != null) {
                headers.put("_correlation_id", message.correlationId().toString());
            }
            if (message.responseAddress() != null) {
                headers.put("_reply_to", message.responseAddress().toString());
            }
            transport.send(message.body(), headers, message.contentType());
            dispatchObservation(message, startedAt, null);
            return CompletableFuture.completedFuture(null);
        } catch (Exception failure) {
            dispatchObservation(message, startedAt, failure);
            return CompletableFuture.failedFuture(failure);
        }
    }

    private void dispatchObservation(OutboxMessage message, long startedAt, Throwable failure) {
        if (hooks.isEmpty()) {
            return;
        }

        String successKind;
        String failureKind;
        switch (message.intent()) {
            case PUBLISH -> {
                successKind = "published";
                failureKind = "publish_faulted";
            }
            case FAULT -> {
                successKind = "fault_published";
                failureKind = "fault_publish_faulted";
            }
            default -> {
                successKind = "sent";
                failureKind = "send_faulted";
            }
        }
        String messageUrn = message.messageTypes().get(0);
        MessageOperationHookEvent event = new MessageOperationHookEvent(
                Instant.now(),
                failure == null ? successKind : failureKind,
                failure == null,
                displayMessageType(messageUrn),
                messageUrn,
                null,
                message.destinationAddress().toString(),
                (System.nanoTime() - startedAt) / 1_000_000.0,
                failure == null ? null : failure.getClass().getName(),
                failure == null ? null : failure.getMessage(),
                message.correlationId() == null ? null : message.correlationId().toString(),
                message.conversationId() == null ? null : message.conversationId().toString(),
                null,
                null,
                null,
                null,
                message.messageId().toString(),
                message.causationMessageId() == null ? null : message.causationMessageId().toString());
        for (BusHook hook : hooks) {
            try {
                hook.handle(event);
            } catch (RuntimeException ignored) {
                // Monitoring hooks must not affect persisted message delivery.
            }
        }
    }

    private static String displayMessageType(String messageUrn) {
        String prefix = "urn:message:";
        return messageUrn.startsWith(prefix)
                ? messageUrn.substring(prefix.length()).replace(':', '.')
                : messageUrn;
    }
}

File: src/MyServiceBus/Persistence/TransportOutboxDispatcher.cs
using System.Diagnostics;
using MyServiceBus.Serialization;

namespace MyServiceBus.Persistence;

public sealed class TransportOutboxDispatcher : IOutboxTransportDispatcher
{
    private readonly ITransportFactory transportFactory;
    private readonly IBusHookDispatcher? hooks;

    public TransportOutboxDispatcher(ITransportFactory transportFactory)
        : this(transportFactory, null)
    {
    }

    public TransportOutboxDispatcher(ITransportFactory transportFactory, IBusHookDispatcher? hooks)
    {
        this.transportFactory = transportFactory ?? throw new ArgumentNullException(nameof(transportFactory));
        this.hooks = hooks;
    }

    public async Task DispatchAsync(OutboxMessage message, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(message);
        var transport = await transportFactory.GetSendTransport(message.DestinationAddress, cancellationToken);
        var serializer = new PersistedEnvelopeSerializer(message.ContentType, message.Body);
        var context = new SendContext([typeof(PersistedEnvelope)], serializer, cancellationToken)
        {
            MessageId = message.MessageId.ToString(),
            RequestId = message.RequestId,
            CorrelationId = message.CorrelationId?.ToString(),
            ConversationId = message.ConversationId,
            InitiatorId = message.InitiatorId,
            CausationMessageId = message.CausationMessageId,
            Intent = MapIntent(message.Intent),
            DestinationAddress = message.DestinationAddress,
            ResponseAddress = message.ResponseAddress,
            FaultAddress = message.FaultAddress
        };

        foreach (var (key, value) in message.Headers)
            context.Headers[key] = value;
        context.Headers["_content_type"] = message.ContentType;
        context.Headers["_message_id"] = message.MessageId.ToString();
        if (message.CorrelationId is { } correlationId)
            context.Headers["_correlation_id"] = correlationId.ToString();
        if (message.ResponseAddress is { } responseAddress)
            context.Headers["_reply_to"] = responseAddress.ToString();

        var startedAt = Stopwatch.GetTimestamp();
        try
        {
            await transport.Send(PersistedEnvelope.Instance, context, cancellationToken);
            DispatchObservation(message, Stopwatch.GetElapsedTime(startedAt));
        }
        catch (Exception exception)
        {
            DispatchObservation(message, Stopwatch.GetElapsedTime(startedAt), exception);
            throw;
        }
    }

    private void DispatchObservation(OutboxMessage message, TimeSpan duration, Exception? exception = null)
    {
        if (hooks?.IsEnabled != true)
            return;

        var (successKind, failureKind) = message.Intent switch
        {
            OutboxDeliveryIntent.Publish => ("published", "publish_faulted"),
            OutboxDeliveryIntent.Fault => ("fault_published", "fault_publish_faulted"),
            _ => ("sent", "send_faulted")
        };
        var messageUrn = message.MessageTypes[0];
        hooks.Dispatch(MessageOperationHookEvent.Create(
            exception is null ? successKind : failureKind,
            exception is null,
            DisplayMessageType(messageUrn),
            messageUrn,
            null,
            message.DestinationAddress.ToString(),
            duration,
            exception,
            message.CorrelationId?.ToString(),
            message.ConversationId?.ToString(),
            messageId: message.MessageId.ToString(),
            causationMessageId: message.CausationMessageId?.ToString(),
            requestId: message.RequestId?.ToString(),
            responseAddress: message.ResponseAddress?.ToString(),
            messageIntent: MapIntent(message.Intent).ToString()));
    }

    private static string DisplayMessageType(string messageUrn)
    {
        const string prefix = "urn:message:";
        return messageUrn.StartsWith(prefix, StringComparison.Ordinal)
            ? messageUrn[prefix.Length..].Replace(':', '.')
            : messageUrn;
    }

    private static MessageIntent MapIntent(OutboxDeliveryIntent intent) => intent switch
    {
        OutboxDeliveryIntent.Send => MessageIntent.Send,
        OutboxDeliveryIntent.Publish => MessageIntent.Publish,
        OutboxDeliveryIntent.Reply => MessageIntent.Reply,
        OutboxDeliveryIntent.Fault => MessageIntent.Publish,
        _ => throw new ArgumentOutOfRangeException(nameof(intent), intent, "Unsupported outbox delivery intent.")
    };

    private sealed class PersistedEnvelope
    {
        public static PersistedEnvelope Instance { get; } = new();
    }

    private sealed class PersistedEnvelopeSerializer(string contentType, ReadOnlyMemory<byte> body) : IMessageSerializer
    {
        public string ContentType { get; } = contentType;

        public MessageBody GetMessageBody<T>(MessageSerializationContext<T> context) where T : class =>
            new ByteArrayMessageBody(body.ToArray());
    }
}
